File: primitives/primitives.py
"""
PRIMITIVE DATA TYPES IN PYTHON
==============================

In Python, everything is an object, but some types work like primitives in other languages.
There are 4 main primitive (built-in) data types:
1. Integer (int)
2. Float (float)
3. String (str)
4. Boolean (bool)

There is also NoneType (complex/list/dict/set/tuple are compound/container types).

All keys properties:
- Immutable (int, float, bool, str, tuple)
- Dynamic Typing (no need to declare type)
"""

# ==========================================
# 1. Integer (int)
# ==========================================
print("--- 1. Integer ---")
# Basics:
age = 25
count = 100

# Advanced: Arbitrary Precision
# Python integers have arbitrary precision (limited only by memory)
big_num = 123456789012345678901234567890 * 1234567890
print(f"Big Number: {big_num}")

# Division behavior
# print(10 / 0)  # ZeroDivisionError: division by zero
print(10 / 3)  # 3.3333... (float division)
print(10 // 3) # 3 (floor division)

# Binary, Octal, Hex representations
print(0b1010) # Binary (10)
print(0o12)   # Octal (10)
print(0xA)    # Hex (10)

# ==========================================
# 2. Float (float)
# ==========================================
print("\n--- 2. Float ---")
# Basics:
price = 19.99
scientific = 1.2e3 # 1200.0

# Advanced: Precision Issues
# Floating point arithmetic based on IEEE 754
print(0.1 + 0.2) # 0.30000000000000004
# Use math.isclose() or decimal module for high precision
import math
print(math.isclose(0.1 + 0.2, 0.3)) # True

# Infinity and NaN
inf = float('inf')
nan = float('nan')
print(inf > 1000000000) # True

# ==========================================
# 3. String (str)
# ==========================================
print("\n--- 3. String ---")
# Basics:
name = "John"
single_quote = 'Hello' # Single quotes are same as double quotes in Python

# Multiline Strings (using triple quotes)
multi = """This is a
multiline string"""
print(multi)

# Advanced: Immutability
s = "hello"
# s[0] = 'H' # TypeError: 'str' object does not support item assignment
s = "Hello" # Reassignment is fine
print(s)

# Formatting (f-strings - Python 3.6+)
print(f"Name: {name}, Age: {age}") # Moved to template example above

# Raw Strings (useful for regex/paths)
path = r"C:\Users\Name" 
print(path)

# Slicing [start:stop:step]
text = "Python Programming"
print(text[0:6])   # 'Python'
print(text[::-1])  # 'gnimmargorP nohtyP' (Reverse)
print(text[::2])  # 'Pto rgamn' (Every 2nd character)

# ==========================================
# 4. Boolean (bool)
# ==========================================
print("\n--- 4. Boolean ---")
# Basics:
is_active = True # Capital T
is_deleted = False # Capital F

# Advanced: Boolean is a subclass of int
print(int(True))  # 1
print(int(False)) # 0
print(True + True) # 2 (valid code, but bad practice)

# Falsy values in Python:
# False, 0, 0.0, "", [], (), {}, None
if []:
    print("Won't print")
else:
    print("Empty list is Falsy")

# ==========================================
# 5. None (NoneType)
# ==========================================
print("\n--- 5. None ---")
# Represents the absence of a value (similar to null in other languages)
x = None
print(x) # None

# Identity check
# Always use 'is' to check for None, not '=='
print(x is None) # True
y = [] # Empty list is falsy but not None
print(y is None) # False

# ==========================================
# 6. Complex Numbers
# ==========================================
print("\n--- 6. Complex ---")
# Python has built-in support for complex numbers
z = 2 + 3j
print(z.real) # 2.0
print(z.imag) # 3.0

# ==========================================
# 7. Decimal Module
# ==========================================
print("\n--- 7. Decimal Module ---")
from decimal import Decimal, getcontext

# Why use Decimal?
# Standard floats have precision issues:
print(f"Float: {0.1 + 0.2}") # 0.30000000000000004

# Decimal fixes this (IMPORTANT: pass values as strings to avoid float conversion errors first)
d1 = Decimal('0.1')
d2 = Decimal('0.2')
print(f"Decimal: {d1 + d2}") # 0.3

# Controlling Precision
getcontext().prec = 6
print(Decimal('1') / Decimal('7')) # 0.142857

File: primitives/primitives.js
/**
 * PRIMITIVE DATA TYPES IN JAVASCRIPT
 * ==================================
 * 
 * In JavaScript, a primitive (primitive value, primitive data type) is data that is not an object and has no methods.
 * There are 7 primitive data types:
 * 1. string
 * 2. number
 * 3. bigint
 * 4. boolean
 * 5. undefined
 * 6. symbol
 * 7. null
 * 
 * All primitives are immutable, meaning they cannot be altered.
 */

// ==========================================
// 1. String
// ==========================================
console.log("--- 1. String ---");
// Basics:
let name = "John";
let singleQuote = 'Hello'; // Single quotes are also fine

// Template Literals (Backticks) - allow embedded expressions
let template = `Welcome, ${name}`;
console.log(template);

// Multiline Strings (using backticks)
let multi = `This is a
multiline string`;
console.log(multi);

// Advanced: Immutability
let str = "Hello";
// str[0] = "h"; // This does nothing (in strict mode, it might throw error)
console.log(str); // Output: "Hello"
// However, the variable can be reassigned
str = "World";
console.log(str); // Output: "World"

// ==========================================
// 2. Number
// ==========================================
console.log("\n--- 2. Number ---");
// Basics:
let age = 25;
let price = 19.99;

// Advanced: Special numeric values
let infinity = Infinity;
let negativeInfinity = -Infinity;
let notANumber = NaN; // Result of invalid math operation

console.log(1 / 0); // Infinity
console.log("text" / 2); // NaN

// Floating point precision issue (IEEE 754)
console.log(0.1 + 0.2); // 0.30000000000000004
console.log(0.1 + 0.2 === 0.3); // false

// range
console.log(Number.MAX_SAFE_INTEGER); // 2^53 - 1

// Binary, Octal, Hex representations
console.log(0b1010); // Binary (10)
console.log(0o12);   // Octal (10)
console.log(0xA);    // Hex (10)

// ==========================================
// 3. BigInt
// ==========================================
console.log("\n--- 3. BigInt ---");
// Created by appending 'n' to the end of an integer
// Used for numbers larger than Number.MAX_SAFE_INTEGER
// let bigNumber = 1234567890123456789012345678901234567890n; // Syntax depends on environment support
console.log(0n);     // BigInt (0)
let anotherBig = BigInt("1234567890123456789012345678901234567890");

console.log(anotherBig);
// Note: You cannot mix BigInt and other types in operations
// console.log(1n + 2); // TypeError

// ==========================================
// 4. Boolean
// ==========================================
console.log("\n--- 4. Boolean ---");
let isActive = true;
let isDeleted = false;

// Advanced: Truthy and Falsy values
// Falsy values: false, 0, -0, 0.0, 0n, "", null, undefined, NaN
if (0) {
    console.log("This won't run");
} else {
    console.log("0 is falsy");
}

// ==========================================
// 5. Undefined
// ==========================================
console.log("\n--- 5. Undefined ---");
// A variable that has not been assigned a value is of type undefined.
let x;
console.log(x); // undefined
console.log(typeof x); // "undefined"

// ==========================================
// 6. Null
// ==========================================
console.log("\n--- 6. Null ---");
// Represents the intentional absence of any object value.
let y = null;
console.log(y); // null

// Advanced: typeof null bug
console.log(typeof null); // "object" (This is a historical bug in JS)

// Difference between null and undefined
// undefined means a variable has been declared but has not yet been assigned a value
// null is an assignment value. It can be assigned to a variable as a representation of no value
console.log(null == undefined); // true
console.log(null === undefined); // false

// ==========================================
// 7. Symbol
// ==========================================
console.log("\n--- 7. Symbol ---");
// A "Symbol" represents a unique identifier.
let id1 = Symbol("id");
let id2 = Symbol("id");

console.log(id1 === id2); // false - Symbols are always unique, even with same description

// Use case: Hidden object properties
let user = {
    name: "John"
};
let id = Symbol("id");
user[id] = 12345;
console.log(user[id]); // 12345

// Symbols are ignored in for...in loops
for (let key in user) {
    console.log(key); // only "name"
}
